src/coreComponents/common/Enum.hpp

```diff
@@ -0,0 +1,217 @@
+ */
+
+/**
+ * @file Enum.hpp
+ *
+ * Collection of utilities to facilitate I/O of enumeration types.
+ * Provides a macro definition that allows associating string names
+ * with enumeration constants and a set of functions that make use
+ * of these strings, like stream insertion/extraction operators.
+ */
+
+#ifndef GEOSX_COMMON_ENUM_HPP_
+#define GEOSX_COMMON_ENUM_HPP_
+
+#include "codingUtilities/StringUtilities.hpp"
+#include "common/DataTypes.hpp"
+#include "common/Logger.hpp"
+
+#include <iostream>
+#include <sstream>
+#include <array>
+#include <type_traits>
+#include <algorithm>
+
+namespace geosx
+{
+
+namespace internal
+{
+
+/**
+ * @brief Simple compile-time variadic function that counts its arguments.
+ * @tparam ARGS variadic pack of argument types
+ * @return the number of arguments passed
+ */
+template< typename ... ARGS >
+constexpr int countArgs( ARGS ... )
+{
+  return sizeof...( ARGS );
+}
+
+/**
+ * @brief Provides some useful properties (like string representation) for enumerations.
+ * @tparam T the enumeration type
+ * @note This template is meant to be specialized using ENUM_STRINGS macro.
+ */
+template< typename T >
+struct EnumStringsBase
+{
+  /// Flag used to distinguish enumeration types that have strings.
+  static constexpr bool defined = false;
+};
+
+}
+
+/**
+ * @brief Associate a list of string names with enumeration values.
+ * @param ENUM the enumeration type
+ * @param ... list of names (C-string literals)
+ *
+ * Conditions (not enforced but won't work correctly if violated):
+ *  - the macro must be called in geosx namespace (not any nested namespace therein)
+ *  - the number and order of string arguments passed must match the number of enum values
+ *  - enumeration constants must not have custom values assigned
+ *
+ * After the macro has been called, template instantiation EnumStrings<ENUM>
+ * may be used to get access to strings at runtime. While not strictly necessary,
+ * it is recommended that macro call immediately follows the enum definition.
+ */
+#define ENUM_STRINGS( ENUM, ... )                                     \
+namespace internal                                                    \
+{                                                                     \
+template<>                                                            \
+struct EnumStringsBase< ENUM >                                        \
+{                                                                     \
+  static_assert( std::is_enum< ENUM >::value, "Not an enumeration" ); \
+  static constexpr bool defined = true;                               \
+                                                                      \
+  static auto const & get()                                           \
+  {                                                                   \
+    constexpr int N = internal::countArgs( __VA_ARGS__ );             \
+    static constexpr std::array< char const *, N > ss{ __VA_ARGS__ }; \
```

Comment:
Storing strings as local static array avoids all kinds of linking issues.

Reply:
nice!